### ISSUE TITLE
Crash when parsing mysql query with union of subquery having group/order by and used in join

### ISSUE THREAD

**Issue by jitsumi** [sqlfluff/sqlfluff#731]
## Expected Behaviour
Parse correctly the valid mysql query
(check https://dev.mysql.com/doc/refman/8.0/en/union.html section `ORDER BY and LIMIT in Unions`)

## Observed Behaviour
Crash with command `sqlfluff parse --dialect mysql file.sql`:
```
Traceback (most recent call last):
  File "/usr/local/bin/sqlfluff", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/cli/commands.py", line 249, in fix
    result = lnt.lint_paths(paths, verbosity=verbose, fix=True, ignore_non_existent_files=False)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/linter.py", line 912, in lint_paths
    ignore_non_existent_files=ignore_non_existent_files))
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/linter.py", line 898, in lint_path
    fix=fix, config=config))
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/linter.py", line 713, in lint_string
    parsed, vs, time_dict = self.parse_string(s=s, fname=fname, verbosity=verbosity, config=config)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/linter.py", line 647, in parse_string
    parsed = file_segment.parse(parse_context=context)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/segments_base.py", line 459, in parse
    incr='parse_depth', match_depth=0, recurse=True
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/segments_base.py", line 703, in expand
    res = stmt.parse(parse_context=parse_context)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/segments_base.py", line 459, in parse
    incr='parse_depth', match_depth=0, recurse=True
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/segments_base.py", line 703, in expand
    res = stmt.parse(parse_context=parse_context)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/segments_base.py", line 459, in parse
    incr='parse_depth', match_depth=0, recurse=True
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/segments_base.py", line 703, in expand
    res = stmt.parse(parse_context=parse_context)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/segments_base.py", line 412, in parse
    match_segment=self.__class__.__name__
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/grammar.py", line 108, in _match
    m = self.match(segments, parse_context=parse_context)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/grammar.py", line 1114, in match
    mid_seg, parse_context=parse_context.copy(incr='match_depth'))
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/grammar.py", line 108, in _match
    m = self.match(segments, parse_context=parse_context)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/grammar.py", line 1233, in match
    code_only=False)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/grammar.py", line 532, in _bracket_sensitive_look_ahead_match
    code_only=code_only)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/parser/grammar.py", line 356, in _look_ahead_match
    "Segment matched in simple parsing, but failed later. Report this."
ValueError: Segment matched in simple parsing, but failed later. Report this.
```

## Steps to Reproduce
parse the following mysql sql query (with `--dialect mysql`)
`sqlfluff parse --dialect mysql file.sql`
```SQL
SELECT *
FROM t1
         INNER JOIN (
        (SELECT * FROM t2 ORDER BY a LIMIT 10)
        UNION
        (SELECT * FROM t3 ORDER BY a LIMIT 10)
) AS unionname ON t1.id = unionname.id
```

## Version
sqlfluff version: 0.3.6
python: 3.7

## Configuration
Default

**Comment by barrywhart** [comment]
I tested this with the latest code in `master` as of today (commit `79767aae39768b5e9c5b8b6cf50f0a589bc075a4`), and it parses okay. So I believe the issue is fixed.

**Comment by alanmcruickshank** [comment]
Agreed. I believe this is fixed with the now released 0.4.0. I'm going to close this issue. 

@jitsumi - if this query still isn't parsing for you with 0.4.0 then can you reopen this issue?
